README.md
# tello-drone

### Disclaimer:

Im a student learning how to work and publish packages on NPM. This is a simple package that allows you to communicate with a DJI Tello drone in a much easier way. Any feedback / help / improvement is appreaciated and i will hopefully be able to look into fixing bugs and such when i can.

This project is currently a work in progress, but as of 1.0.1 it can communicate with a tello drone with the sample code bellow.

## Changelog
You can see the latest changes in the **CHANGELOG.md** file in the root of this repository

## v3.0.0 Breaking changes
- Package is now only async.
- Removed offsetBuffer and async option from config.
- Removed DELAYS attribute from the tello-data.json and the tello package.

### v3.0.0 General changes
- Replaced large sections of the code that could be replaced with native node modules, such as events and assert.
- Added testing and linting

## Information

The goal of this package is to provide an independent (no dependecies), simple and easy-to-use interface between nodejs and a tello drone. You can communicate with the drone asynchronously. When you connect to the drone it **automatically sends the 'command' command**, so that is not nececary.

## Error handling

The drone gives no feedback on errors. Command verification is already built in, it will throw an error if a command is invalid.
If you recieve back an error and the drone does not execute the command, the most common reason in my experience is low battery.
Charging the battery and trying again solves that error issue.

## Methods and options

Note that all the parameters for a command needs to be an object as a second parameter to the send command. You can reference the [tello documentation here](https://dl-cdn.ryzerobotics.com/downloads/tello/20180910/Tello%20SDK%20Documentation%20EN_1.3.pdf) to what parameters needs to go along with a specific command. Its important that the parameters in the object **needs to be in the same order** as in the tello documentation. For commands that only has a single value (usually x in the tello documentation) you only need to pass in an object such as `{ value: 60 }`.

Below is a general description of all the parameters and events:

```js
const Tello = require("tello-drone");

// All option parameters are optional, default values shown
const drone = new Tello({
      host: "192.168.10.1",     // manually set the host.
      port: "8889",             // manually set the port.
      statePort: "8890",        // manually set the state port.
      skipOK: false,            // dont send the OK message.
})

// Sends a command to the drone
drone.send("battery?")

// Sends a command to the drone without verifying the command
drone.forceSend("battery?")

// Attaches a callback to a specific event
drone.on(event, callback)

/** All events of the on method
 * connection   - Fired when connected to the drone.             callback()
 * state        - Fired when the drone sends its state message.  callback(stateObject, udpConnection)
 * send         - Fired when a command is sent.                  callback(error, messageLength)
 * message      - Fired when the drone sends a status message.   callback(message, udpConnection)
 */

```

# Code example

```js
const Tello = require("tello-drone");

const drone = new Tello();

drone.on("connection", () => {
    console.log("Connected to drone");
});

drone.on("state", state => {
    console.log("Recieved State > ", state);
});

drone.on("send", (err, length) => {
    if (err) console.log(err);

    console.log(`Sent command is ${length} long`);
});

drone.on("message", message => {
    console.log("Recieved Message > ", message);
});

drone.on("connection", async () => {
    try {
        await drone.send("takeoff");
        await drone.send("flip", { value: "f" });
        await drone.send("land");
    } catch (error) {
        console.log(error)
        drone.send("land")
        //A short delay so that the land command can be sent before exiting
        setTimeout(process.exit)
    }
});
```

# Video demo

[![Youtube video, broken :/](http://img.youtube.com/vi/pxh4rlVNd4E/0.jpg)](http://www.youtube.com/watch?v=pxh4rlVNd4E "NPM tello-drone v2.0.0 demo")

## Avaliable commands to send

Tellos official SDK documentation with all commands:

#### Check your firmware version for the right manual (most likely 1.3)

1.0: https://dl-cdn.ryzerobotics.com/downloads/tello/0228/Tello+SDK+Readme.pdf

1.3: https://dl-cdn.ryzerobotics.com/downloads/tello/20180910/Tello%20SDK%20Documentation%20EN_1.3.pdf

2.0: https://dl-cdn.ryzerobotics.com/downloads/Tello/Tello%20SDK%202.0%20User%20Guide.pdf

Here are the most used ones:
**command, takeoff, land, up, down, left, right, forward, back, flip, speed, speed?, battery?**


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [3.0.5] - 2020-05-23

### Changed
- ported the repo from JS to TS

### Added
- type files for the drone and its methods
- this changelog

## [3.0.4] - 2020-03-15
### Changed
- fixed a link in the readme
- updated documentation on the options parameter


## [3.0.3] - 2020-03-15
### Added
- added missing z1 and z2 parameters of the curve command to tello-data.json

## [3.0.2] - 2020-03-15

### Added
- added missing rc command to tello-data.json

## [3.0.1] - 2019-11-27

### Changed
- added breaking changes note on readme


## [3.0.0] - 2019-11-27

### Added
- testing
- linting

### Changed
- removed delays from the tello-data.json file
- removed 'DELAYS' property from index.js
- removed 'bufferOffset' option
- changed the structure of the repository

## [2.1.0] - 2019-01-24

### Changed
- replaced large parts of code with native nodeJS modules (assert, events)
- changed drone function into a class
- removed 'sync' methods, package is now only async

## [2.0.1] - 2019-01-22

### Changed
- fixed bug where class did not send the 'command' command

## [2.0.0] - 2019-01-22

### Added
- added 'async' option to drone constructor
- added method syncSend()
- added method asyncSend()
- added method forceSendSync()
- added method forceSendAsync()

### Changed
- the send() method now handles sync or async depending on options

## [1.0.2] - 2019-01-07
Version patch, nothing changed

## [1.0.1] - 2019-01-07

### Changed
- Fixed spelling in method documentation

## [1.0.0] - 2019-01-07
First publish

lib/classes.js
const dgram = require("dgram");
const EventEmitter = require("events");
const assert = require("assert");

const { verifyCommand, parseDroneState, formatCommand } = require("./utility.js");

module.exports.Drone = class Drone {
    /**
     * Connects to the drone and returns a new Drone object
     * @param {Object} config - The optional config object
     * @param {String} config.host - The IP of the Tello drone
     * @param {String} config.port - The port of the UDP communication
     * @param {String} config.statePort - The port of the UDP state communication
     * @param {Boolean} config.skipOK - Dont send the ok message
     */
    constructor({
        host = "192.168.10.1",
        port = "8889",
        statePort = "8890",
        skipOK = false,
    } = {}) {
        assert.equal(typeof host, "string");
        assert.equal(typeof port, "string");
        assert.equal(typeof statePort, "string");
        assert.equal(typeof skipOK, "boolean");

        this.HOST = host;
        this.MAIN_PORT = port;
        this.STATE_PORT = statePort;
        this.droneIO = dgram.createSocket("udp4");
        this.droneState = dgram.createSocket("udp4");
        this.connected = false;

        this.events = new EventEmitter();

        this.droneIO.bind(this.MAIN_PORT);
        this.droneState.bind(this.STATE_PORT);

        this.droneState.on("message", stateBuffer => (
            this.events.emit("state", parseDroneState(stateBuffer))
        ));

        this.droneIO.on("message", (...args) => {
            let [message] = args;

            message = Buffer.isBuffer(message) ? message.toString() : message;

            if (message !== "ok") {
                return this.events.emit("message", message);
            }

            this.events.emit("_ok");

            if (!this.connected) {
                this.connected = true;
                this.events.emit("connection");
            }

            if (!skipOK) this.events.emit("message", message);
        });

        // Add a minor delay so that the events can be attached first
        setTimeout(this.send.bind(this, "command"));
    }

    /**
     * Sends a command to the drone via UDP
     * @param {
     * "command" | "takeoff" | "land" | "streamon" | "streamoff" | "emergency" | "up" | "down" | "left" | "right" | "forward" |
     * "back" |"cw" | "ccw" | "flip" | "go" | "curve" | "speed" | "rc" | "wifi" | "speed?" | "battery?" | "time?" | "wifi?" | "sdk?" | "sn?"
     * } command
     * @param {Object} options - The option object
     * @param {Number} options.value - Number of cm the drone should move in a certain direction
     * @param {Number} options.speed - How fast the drone should move. Max: 100, Min: 10.
     * @param {Number} options.x - Number of cm the drone should move in the x axis. Max: 500, Min: -500.
     * @param {Number} options.y - Number of cm the drone should move in the y axis. Max: 500, Min: -500.
     * @param {Number} options.z - Number of cm the drone should move in the z axis. Max: 500, Min: -500.
     * @param {Number} options.x1 - Starting x position for the curve command. Max: 500, Min: -500.
     * @param {Number} options.x2 - Ending x position for the curve command. Max: 500, Min: -500.
     * @param {Number} options.y1 - Starting y position for the curve command. Max: 500, Min: -500.
     * @param {Number} options.y2 - Ending y position for the curve command. Max: 500, Min: -500.
     * @param {Number} options.z1 - Starting z position for the curve command. Max: 500, Min: -500.
     * @param {Number} options.z2 - Ending z position for the curve command. Max: 500, Min: -500.
     */
    send(command, options, force) {
        const error = verifyCommand(command, options);
        let formattedCommand = command;

        if (options) {
            formattedCommand = formatCommand(command, options);
        }

        return new Promise((resolve, reject) => {
            if (error && !force) return reject(error);

            this.droneIO.send(
                formattedCommand,
                0,
                formattedCommand.length,
                this.MAIN_PORT,
                this.HOST,
                this.events.emit.bind(this.events, "send"),
            );

            // If the command is a read command resolve immediately
            if (formattedCommand.includes("?")) {
                return resolve();
            }

            // otherwise wait for OK from drone
            this.events.once("_ok", resolve);
        });
    }

    /**
     * Sends a command to the drone via UDP without verifying the command
     * @param {"command" | "takeoff" | "land" | "streamon" | "streamoff" | "emergency" | "up" | "down" | "left" | "right" | "forward" |
     * "back" | "cw" | "ccw" | "flip" | "go" | "curve" | "speed" | "rc" | "wifi" | "speed?" | "battery?" | "time?" | "wifi?" | "sdk?" | "sn?"
     * } command
     * @param {Object} options - The option object
     * @param {Number} options.value - Number of cm the drone should move in a certain direction
     * @param {Number} options.speed - How fast the drone should move. Max: 100, Min: 10.
     * @param {Number} options.x - Number of cm the drone should move in the x axis. Max: 500, Min: -500.
     * @param {Number} options.y - Number of cm the drone should move in the y axis. Max: 500, Min: -500.
     * @param {Number} options.z - Number of cm the drone should move in the z axis. Max: 500, Min: -500.
     * @param {Number} options.x1 - Starting x position for the curve command. Max: 500, Min: -500.
     * @param {Number} options.x2 - Ending x position for the curve command. Max: 500, Min: -500.
     * @param {Number} options.y1 - Starting y position for the curve command. Max: 500, Min: -500.
     * @param {Number} options.y2 - Ending y position for the curve command. Max: 500, Min: -500.
     */
    forceSend(command, options) {
        this.send.bind(this, command, options, true)();
    }

    /**
     * Attaches a callback to a specific event
     *
     * connection   - Fired when connected to the drone.             callback()
     *
     * state        - Fired when the drone sends its state message.  callback(stateObject, udpConnection)
     *
     * send         - Fired when a command is sent.                  callback(error, messageLength)
     *
     * message      - Fired when the drone sends a status message.   callback(message, udpConnection)
     *
     * @param {String} event - Event name
     * @param {Function} callback - Callback fired upon event
     */
    on(event, callback) {
        this.events.on(event, callback);
    }
};
